client/src/pages/EmailSentForm.tsx

```diff
@@ -0,0 +1,107 @@
+import { useState } from 'react';
+import { useLocation, useNavigate } from 'react-router-dom';
+
+export function EmailSentForm() {
+  const location = useLocation();
+  const navigate = useNavigate();
+  const recipe = location.state;
+  const [toAddress, setToAddress] = useState<string>('');
+  const [isValidEmail, setIsValidEmail] = useState<boolean>(true);
+  async function handleSubmit(e: React.FormEvent<HTMLFormElement>) {
+    e.preventDefault();
+    const formData = new FormData(e.currentTarget);
+    const objectData = Object.fromEntries(formData);
+    objectData.fromAddress = 'willzhang73@gmail.com';
+    const body = {
+      recipe,
+      objectData,
+    };
+    try {
+      const req = {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(body),
+      };
+      const response = await fetch('/api/sendemail', req);
+      if (!response) throw new Error('Network response not ok');
+      const data = await response.json();
+      if (data === 'recipe undefined') {
```

Comment:
Usually you would check the payload you are sending for this error, so that you never send an undefined recipe.

That is, after line 18, before the `try`, you would check that `recipe` is defined, and if not, issue the alert and return.